project/iar1000/README.md

```diff
@@ -0,0 +1,98 @@
+
+## System Architecture
+
+### Workflow
+
+The implementation itself is not literally sweeping the sweep line through the whole plane and adapting the beach line on it's way, but skipping to locations where important events take place. This is also called an event-based algorithm. The two major events that have impact on the diagram are the following:
+
+1) **Point Event**: The sweep line encounters a new target point, a new arc is added to the beach line tracking the boarder of the new cell
+2) **Circle Event**: The length of an arc shrinks to zero, a new vertex in the Voronoi diagram is created from where a new cell boundary starts
+
+#### Point Event
+
+The sweep line encounters target point $t_k$. The horizontal line $h_k$ going through $t_k$ is going to intersect with a parabola $p_k \in P$ to the left of the sweep line.
+
+A new parabola is introduced to the beach line at the intersection of $h_k$ and $p_k$. It is infitesimally small in the beginning and starts to grow as the sweep line progresses, tracking the equidistance level curve for the newly incorporated point $t_k$.
+
+As the sweep line progresses only in one direction, we know the order of the point events in advance. Sorting the target points in advance allows us an efficient lookup for the next point event.
+
+#### Circle Event
+
+Circle events are created on the beach line by three neighbouring target points. Each of the three points has an arc associated with it, all part of the beach line, which indicates equidistance from the closest target point and the sweep line. While all of the target points grow their cell, their respective arcs on the beach line grow with them, until one of them is "squashed" to length zero by the other two. This is going to happen at point $v_i$, which has the same distance to all three neighbours. A Voronoi vertex is created at this location and the parabola with empty length is removed, because we know that in the final diagram, a border is starting at that point.
+
+### Pseudo Code
+
+We keep track of the events in a event queue $Q$, handling them one after the other.
+
+```
+insert all Point Events to Q, sorted by x-coordinate
+while Q is not empty:
+    pop event E
+    if E is a Point Event:
+        insert new arc to the beach line
+        check for new Circle Events
+    else:
+        create a new vertex in the diagram D
+        remove the shrunk arc from the beach line
+        delete invalidated events
+        check and add new Circle Events
+```
+
+### Interfaces
+
+The software will have one central datastructure, namely the ``VoronoiDiagram``. The Voronoi diagram $V$ is implemented as a [Doubly Connected Edge List (DCEL)](https://en.wikipedia.org/wiki/Doubly_connected_edge_list), which stores the planar graph as a set of (half-)edges, vertices and cells. It is special in that each edge is represented as two half-edges, each associated to one of the cells forming the edge.
```

Comment:
It is a crucial approach to design numerical software based on data structure.  It's hard to go astray this way.  You can probably make an effective description of the software architecture based on your understandings to the data structure and the numerical methods for the Voronoi diagrams.